internal/handlers/jsonb1/msg_find.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jsonb1
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/common"
+	"github.com/FerretDB/FerretDB/internal/pg"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/wire"
+)
+
+// MsgFind finds documents in a collection or view and returns a cursor to the selected documents.
+func (s *storage) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
+	document, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	if err := common.Unimplemented(document, common.UnimplementedFieldsFindOrCount...); err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	common.Ignored(document, s.l, common.IgnoredFieldsFindOrCount...)
+
+	var filter *types.Document
+	var sql, collection string
+	var args []any
+	var placeholder pg.Placeholder
+
+	m := document.Map()
+	db := m["$db"].(string)
+
+	projectionIn, _ := m["projection"].(*types.Document)
+	projectionSQL, projectionArgs, err := s.projection(projectionIn, &placeholder)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	args = append(args, projectionArgs...)
+
+	collection = m["find"].(string)
+	sql = fmt.Sprintf(`SELECT %s FROM %s`, projectionSQL, pgx.Identifier{db, collection}.Sanitize())
+	filter, _ = m["filter"].(*types.Document)
+	if filter.Len() == 0 {
+		filter, _ = m["projection"].(*types.Document)
```

Comment:
Uh, why? Projection is not the same as a filter. Or do I miss something?

Reply:
No just need to think it up how to apply both filters.